planner/core/rule_column_pruning.go

```diff
@@ -81,6 +81,10 @@ func (p *LogicalProjection) PruneColumns(parentUsedCols []*expression.Column) er
 func (p *LogicalSelection) PruneColumns(parentUsedCols []*expression.Column) error {
 	child := p.children[0]
 	parentUsedCols = expression.ExtractColumnsFromExpressions(parentUsedCols, p.Conditions, nil)
+	if len(parentUsedCols) != 0 {
```

Comment:
This block should be run before `parentUsedCols = expression.ExtractColumnsFromExpressions(...`.

Columns in `p.Conditions` can be pruned in selection.

Reply:
I'm confusing, it seems that after `ExtractColumnsFromExpressions` invoke, the `parentUsedCols` variable will contains the Columns in p.Conditions. And it maybe pruned
in `p.inlineProjection(parentUsedCols)`

Reply:
`p.inlineProjection(parentUsedCols)` is pruning columns __NOT__ in `parentUsedCols`.

Reply:
If call `p.inlineProjection(parentUsedCols)` first.
`p.inlineProjection(parentUsedCols)` will prune columns which in `p.Conditions`

Reply:
Columns in `p.Conditions` but not in `parentUsedCols` can be pruned.

For example:
```sql
mysql> explain select a from (select a, b, count(b) over (partition by a) as cnt from t) tt where b < 10 ;
+--------------------------------+---------+-----------+---------------+-------------------------------------------------------+
| id                             | estRows | task      | access object | operator info                                         |
+--------------------------------+---------+-----------+---------------+-------------------------------------------------------+
| Projection_8                   | 0.80    | root      |               | test.t.a                                              |
| └─Selection_9                  | 0.80    | root      |               | lt(test.t.b, 10)                                      |
|   └─Window_10                  | 1.00    | root      |               | count(test.t.b)->Column#5 over(partition by test.t.a) |
|     └─Sort_13                  | 1.00    | root      |               | test.t.a                                              |
|       └─TableReader_12         | 1.00    | root      |               | data:TableFullScan_11                                 |
|         └─TableFullScan_11     | 1.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo                        |
+--------------------------------+---------+-----------+---------------+-------------------------------------------------------+
6 rows in set (0.00 sec)
```

Note the `select a from ... where b > 0`. `Projection_8` is projecting column `a`.

The `inline projection` of Selection actually replaces the role of `Projection_8`, i.e, projecting column `a` ONLY.

Reply:
If prune columns in `p.Conditions`, it seems that the cascades optimizer will do `EliminateProjection` to eliminate the projection. 
The default optimizer  use `projectionEliminator` which has an parameter `canEliminate`.
If root plan is `LogicalProjection`, it will not be eliminate

Reply:
Yes, you are right.

I am not sure what is your question now. Shall we discuss further in [slack](https://join.slack.com/t/tidbcommunity/shared_invite/zt-9vpzdqh2-8LsybcK0US_nqwvfAjSU5A)? Look for _PingYu_ there.